File: 프로그래머스/2/12923. 숫자 블록/숫자 블록.py
def solution(begin, end):
    answer = []
    
    for i in range(begin, end + 1):
        if i == 1:
            answer.append(0)
            continue
        
        max_block = 1  # i의 약수 중 가장 큰 블록 번호
        
        # i의 약수 탐색 (1부터 sqrt(i)까지)
        for j in range(2, int(i**0.5) + 1):
            if i % j == 0:  # j가 약수일 때
                other = i // j  # j에 대응하는 약수
                
                # 10,000,000 이하의 약수 중 큰 값을 찾음
                if other <= 10**7:
                    max_block = max(max_block, other)
                if j <= 10**7:
                    max_block = max(max_block, j)
        
        answer.append(max_block)
    
    return answer

File: 프로그래머스/2/12923. 숫자 블록/숫자 블록.java
import java.util.*;

class Solution {
    private static final int LIMIT = 10_000_000;

    public int[] solution(long begin, long end) {
        int n = (int) (end - begin + 1);
        int[] ans = new int[n];

        for (int i = 0; i < n; i++) {
            long x = begin + i;
            ans[i] = blockNumber(x);
        }
        return ans;
    }

    private int blockNumber(long x) {
        if (x == 1) return 0; 

        int best = 1;
        long root = (long) Math.sqrt(x);

        for (long a = 2; a <= root; a++) {
            if (x % a != 0) continue;

            long b = x / a; 
            if (b <= LIMIT) return (int) b;

            if (a <= LIMIT) best = Math.max(best, (int) a);
        }

        return best;
    }
}